docs/source/general/data.rst

```diff
@@ -2,11 +2,371 @@
+
+*******************************************
+How to use out-of-the-box flashdatamodules
+*******************************************
+
+Flash provides several DataModules with helpers functions.
+Checkout the ``Task Section`` to learn more about them.
+
+********************************
+Why Preprocess and PostProcess ?
+********************************
+
+Currently, it is common practice to implement a :class:`~torch.data.utils.Dataset` and provide them to a :class:`~torch.data.utils.DataLoader`.
+
+However, once the model is trained, lot of engineering work is required to enable the model
+to perform predictions on ``un-processed data`` (called raw data) or to study data transforms impact during training.
+
+But more importantly, it is hard to make a trainer model ready for production.
+
+The :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` have been created to resolve those issues.
+By providing a series of hooks which can overridden with custom data processing logic, the user have a much more granular control on the data processing flow.
+
+But it also makes your code more readable, modular and easy to extend.
+
+To change the processing behavior only on specific stages for a given hook,
+you can prefix all the above hooks adding ``train``, ``val``, ``test`` or ``predict``.
+
+.. note::
+
+    [WIP] We are currently working on a new feature to convert fully implemented :class:`~flash.data.process.Preprocess`
+
+    and :class:`~flash.data.process.Postprocess` are fully implemented, to be fully deployable for ``Endpoints``
+
+    or ``BatchTransformJob`` directly from checkpoints. Stay tuned !
+
+*************************************
+How to customize existing datamodules
+*************************************
+
+Currently, Flash Tasks are implementing using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess`.
+However, it is not a hard requirement and one can still use :class:`~torch.data.utils.Dataset`, but we highly recommend
+using :class:`~flash.data.process.Preprocess` and :class:`~flash.data.process.Postprocess` instead.
+
+Example::
+
+    from flash.data.data_module import DataModule
+
+    dm = DataModule(train_dataset=MyDataset(train=True))
+    trainer = Trainer(fast_dev_run=True)
+    trainer.fit(model, data_module=dm)
+
+In order to customize Flash to your need, you need to know what are :class:`~flash.data.data_module.DataModule`
+and :class:`~flash.data.process.Preprocess` responsibilities.
+
+.. note::
+
+    From this point, we strongly encourage the readers to quickly check the :class:`~flash.data.process.Preprocess` API before getting further.
+
+The :class:`~flash.data.data_module.DataModule` provides ``classmethod`` helpers to build
+:class:`~flash.data.process.Preprocess` and :class:`~flash.data.data_pipeline.DataPipeline`,
+generate ``AutoDataset`` and populate DataLoaders from them.
```

Comment:
What does it mean to generate AutoDataset?

Reply:
AutoDataset are internals class. Tried to clarify this.